Test JSON top-level maps and add documentation about serialization formats (#2796)
confluentinc/ksql: docs/developer-guide/serialization.rst

.. _ksql_serialization:

KSQL Serialization
==================

KSQL currently supports three serialization formats:

*. ``DELIMITED`` supports comma separated values. See :ref:`delimited_format` below.
*. ``JSON`` supports JSON values. See :ref:`json_format` below.
*. ``AVRO`` supports AVRO serialized values. See :ref:`avro_format` below.

.. _delimited_format

---------
DELIMITED
---------

The ``DELIMITED`` format supports comma separated values.

The serialized object should be a Kafka serialized string, which will be split into columns.

For example, given a KSQL statement such as:

.. code:: sql

    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='DELIMITED', ...);

> Then KSQL would split a value of ``120, bob, 49`` into the three fields with ``ID`` of ``120``,

Reviewer: ```suggestion
KSQL splits a value of ``120, bob, 49`` into the three fields with ``ID`` of ``120``,
```

``NAME`` of ``bob`` and ``AGE`` of ``49``.

This data format supports all KSQL data types except ``ARRAY``, ``MAP`` and ``STRUCT``. See
:ref:`data-types` for more details.

.. _json_format

----
JSON
----

The ``JSON`` format supports JSON values.

> The JSON format supports all of KSQL's data types. (See :ref:`data-types` for details). Though,

Reviewer: ```suggestion
The JSON format supports all of KSQL's ref:`data types <data-types>`. Because
```

as JSON itself does not support a map type, KSQL serializes ``MAP``s as JSON objects.  Because of

> this the JSON format can only support ``MAP``s with ``STRING`` keys.

Reviewer: ```suggestion
this the JSON format can only support ``MAP`` objects that have ``STRING`` keys.
```

> The serialized object should be a Kafka serialized string containing a valid JSON value. The foramt

Reviewer: ```suggestion
The serialized object should be a Kafka-serialized string containing a valid JSON value. The format
```

supports JSON objects and top-level primitives, arrays and maps. See below for more info.

JSON Objects
------------

Values that are JSON objects are probably the most common.

For example, given a KSQL statement such as:

.. code:: sql

    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='JSON', ...);

And a JSON value of:

.. code:: json

       {
         "id": 120,
         "name": "bob",
         "age": "49"
       }

> KSQL would deserialize the JSON object's fields into the corresponding field of the stream.

Reviewer: ```suggestion
KSQL deserializes the JSON object's fields into the corresponding fields of the stream.
```

Top-level Primitives
--------------------

> The JSON format supports reading top-level JSON primitives. KSQL can only do so if the target

Reviewer: ```suggestion
The JSON format supports reading top-level JSON primitives, but only if the target
```

> schema contains only a single field of a compatible type.

Reviewer: ```suggestion
schema contains a single field of a compatible type.
```

For example, given a KSQL statement with only a single field in the value schema:

.. code:: sql

    CREATE STREAM x (ID BIGINT) WITH (VALUE_FORMAT='JSON', ...);

And a JSON value of:

.. code:: json

>        10

Reviewer 2: How do we distinguish between 

```JSON
10
```
and 
```JSON
{ "ID":"10"}
```
I assumed we were going to add new formats, `JSON_PRIMITIVE` and `AVRO_PIRIMITIVE`. Or maybe something like `JSON_RAW` and `AVRO_RAW` since we can have ARRAY or MAP too.
This would indicate that the value is not a JSON document or Avro recored.

Author: I distinguish between the two by virtue of the fact they are different ;)

`FORMAT='JSON'` is enough on the deserialize side as you have the data. If the value deserializes to a primitive, then we treat it as a primitive. If we deserial to a JSON object or Avro record then we treat it as such.

> KSQL would deserialize the JSON primitive ``10`` into the ``ID`` field of the stream.

Reviewer: ```suggestion
KSQL deserializes the JSON primitive ``10`` into the ``ID`` field of the stream.
```

However, if the value schema contained multiple fields, for example:

.. code:: sql

    CREATE STREAM x (ID BIGINT, NAME STRING) WITH (VALUE_FORMAT='JSON', ...);

> Deserialization would fail as it is ambiguous as to which field the primitive value should be

Reviewer: ```suggestion
Deserialization fails, because it's ambiguous as to which field the primitive value should be
```

deserialized into.

Top-level Arrays
----------------

> The JSON format supports reading top-level JSON arrays. KSQL can only do so if the target schema

Reviewer: ```suggestion
The JSON format supports reading top-level JSON arrays, but only if the target schema
```

> contains only a single field of a compatible type.

Reviewer: ```suggestion
contains a single field of a compatible type.
```

For example, given a KSQL statement with only a single array field in the value schema:

.. code:: sql

    CREATE STREAM x (REGIONS ARRAY<STRING>) WITH (VALUE_FORMAT='JSON', ...);

And a JSON value of:

.. code:: json

       [
          "US",
          "EMEA"
       ]

> KSQL would deserialize the JSON array into the ``REGIONS`` field of the stream.

Reviewer: ```suggestion
KSQL deserializes the JSON array into the ``REGIONS`` field of the stream.
```

> However, if the value schema contained multiple fields, for example:

Reviewer: ```suggestion
However, if the value schema contains multiple fields, for example:
```

.. code:: sql

    CREATE STREAM x (REGIONS ARRAY<STRING>, NAME STRING) WITH (VALUE_FORMAT='JSON', ...);

> Deserialization would fail as it is ambiguous as to which field the primitive value should be

Reviewer: ```suggestion
Deserialization fails, because it's ambiguous as to which field the primitive value should be
```

deserialized into.

Top-level Maps
--------------

> .. tip:: Care must be take when deserializing JSON objects into a single ``MAP`` field to ensure

Reviewer: ```suggestion
.. tip:: When you deserialize JSON objects into a single ``MAP`` field, be sure that
```

>          the name of the field within the KSQL statement never clashes with any of the keys within

Reviewer: ```suggestion
         the name of the field within the KSQL statement doesn't clash with any of the keys in
```

>          the map.  Any clash can lead to undesirable deserialization artifacts as KSQL will

Reviewer: ```suggestion
         the map.  Any conflicts can lead to undesirable deserialization artifacts, because KSQL
```

>          treat the value as a normal JSON object, not as a map.

Reviewer: ```suggestion
         treats the value as a normal JSON object, not as a map.
```

> The JSON format supports reading a JSON object as a ``MAP``. KSQL can only do so if the target

Reviewer: ```suggestion
The JSON format supports reading a JSON object as a ``MAP``, but only if the target
```

> schema contains only a single field of a compatible type.

Reviewer: ```suggestion
schema contains a single field of a compatible type.
```

For example, given a KSQL statement with only a single map field in the value schema:

.. code:: sql

    CREATE STREAM x (PROPS MAP<STRING, STRING>) WITH (VALUE_FORMAT='JSON', ...);

And a JSON value of:

.. code:: json

       {
          "nodeCount": 10,
          "region": "us-12",
          "userId": "peter"
       }

> KSQL would deserialize the JSON map into the ``PROPS`` field of the stream.

Reviewer: ```suggestion
KSQL deserializes the JSON map into the ``PROPS`` field of the stream.
```

However, if the value schema contained multiple fields, for example:

.. code:: sql

    CREATE STREAM x (PROPS MAP<STRING, STRING>, NAME STRING) WITH (VALUE_FORMAT='JSON', ...);

> Deserialization would fail as it is ambiguous as to which field the primitive value should be

Reviewer: ```suggestion
Deserialization fails, because it's ambiguous as to which field the primitive value should be
```

deserialized into.

> A further potential ambiguity exists when working with top-level maps should any of the keys of the

Reviewer: ```suggestion
A further potential ambiguity exists when working with top-level maps, when any of the keys of the
```

value match the name of the singular field in the target schema.

For example, given:

.. code:: sql

    CREATE STREAM x (PROPS MAP<STRING, STRING>) WITH (VALUE_FORMAT='JSON', ...);

And a JSON value of:

.. code:: json

       {
          "props": {
             "x": "y"
          },
          "region": "us-12",
          "userId": "peter"
       }

> It is now ambiguous to KSQL as to how to deserialize the value: top level map or object? KSQL will

Reviewer: ```suggestion
Deserializing the value is ambiguous: does KSQL deserialize to a top-level map or object? KSQL 
```

> deserialize the value as JSON object, meaning ``PROPS`` will be populated with an entry ``x -> y``

Reviewer: ```suggestion
deserializes the value as a JSON object, meaning ``PROPS`` is populated with an entry ``x -> y``
```

> only.  Such ambiguity can be avoided by ensuring the name of the field using in the KSQL statement

Reviewer: ```suggestion
only. Avoid this kind of ambiguity by ensuring the name of the field in the KSQL statement
```

> never clashes with a property name within the json object.

Reviewer: ```suggestion
never clashes with a property name within the JSON object.
```

.. _avro_format

----
Avro
----

The ``AVRO`` format supports Avro binary serialized records and top-level primitives, arrays and
maps.

The format requires KSQL to be configured to store and retrieve the Avro schemas from the |sr-long|.

> See :ref:`install_ksql-avro-schema` for more info.

Reviewer: ```suggestion
For more information, see :ref:`install_ksql-avro-schema`.
```

------------
Avro Records
------------

Avro records can be deserialized into matching KSQL schemas.

For example, given a KSQL statement such as:

.. code:: sql

    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='JSON', ...);

> And a Avro record serialized with the schema:

Reviewer: ```suggestion
And an Avro record serialized with the schema:
```

.. code:: json

       {
         "type": "record",
         "namespace": "com.acme",
         "name": "UserDetails",
         "fields": [
           { "name": "id", "type": "long" },
           { "name": "name", "type": "string" }
           { "name": "age", "type": "int" }
         ]
       }

> KSQL would deserialize the Avro record's fields into the corresponding field of the stream.

Reviewer: ```suggestion
KSQL deserializes the Avro record's fields into the corresponding fields of the stream.
```

-------------------------------------
Top-level primitives, arrays and maps
-------------------------------------

> The Avro format supports reading top-level primitives, arrays and maps. KSQL can only do so if the

Reviewer: ```suggestion
The Avro format supports reading top-level primitives, arrays and maps, but only if the
```

> target schema contains only a single field of a compatible type.

Reviewer: ```suggestion
target schema contains a single field of a compatible type.
```

For example, given a KSQL statement with only a single field in the value schema:

.. code:: sql

    CREATE STREAM x (ID BIGINT) WITH (VALUE_FORMAT='JSON', ...);

And a Avro value serialized with the schema:

.. code:: json

       {
         { "type": "long" }
       }

KSQL can deserialize the values into the ``ID`` field of the stream.

However, if the value schema contained multiple fields, for example:

.. code:: sql

    CREATE STREAM x (ID BIGINT, NAME STRING) WITH (VALUE_FORMAT='JSON', ...);

> Deserialization would fail as it is ambiguous as to which field the primitive value should be

Reviewer: ```suggestion
Deserialization fails, because it's ambiguous as to which field the primitive value should be
```

KSQL Serialization
==================

KSQL currently supports three serialization formats:

*. ``DELIMITED`` supports comma separated values. See :ref:`delimited_format` below.
*. ``JSON`` supports JSON values. See :ref:`json_format` below.
*. ``AVRO`` supports AVRO serialized values. See :ref:`avro_format` below.

.. _delimited_format

---------
DELIMITED
---------

The ``DELIMITED`` format supports comma separated values.

The serialized object should be a Kafka-serialized string, which will be split into columns.

For example, given a KSQL statement such as:

.. code:: sql

    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='DELIMITED', ...);

KSQL splits a value of ``120, bob, 49`` into the three fields with ``ID`` of ``120``,
``NAME`` of ``bob`` and ``AGE`` of ``49``.

This data format supports all KSQL :ref:`data types <data-types>` except ``ARRAY``, ``MAP`` and
``STRUCT``.

.. _json_format

----
JSON
----

The ``JSON`` format supports JSON values.

The JSON format supports all of KSQL's ref:`data types <data-types>`. As JSON does not itself
support a map type, KSQL serializes ``MAP``s as JSON objects.  Because of this the JSON format can

> only support ``MAP`` objects that have ``STRING`` keys.

Author: For sure, but not in this PR as you say.
